internal/vault/vault.go

```diff
@@ -375,6 +388,79 @@ func (v *Vault) requestTokenWithAppRoleRef(client Client, appRole *v1.VaultAppRo
 	return token, nil
 }
 
+func (v *Vault) requestTokenWithClientCertificate(client Client, clientCertificateAuth *v1.VaultClientCertificateAuth) (string, error) {
+	// If secretName is set, load client certificate from Secret, otherwise assume that a
+	// fitting client certificate is loaded in the client already.
+	if len(clientCertificateAuth.SecretName) != 0 {
+		secret, err := v.secretsLister.Secrets(v.namespace).Get(clientCertificateAuth.SecretName)
+		if err != nil {
+			return "", err
+		}
+
+		cert, ok := secret.Data["tls.crt"]
+		if !ok {
+			return "", fmt.Errorf("no data for tls.crt in secret '%s/%s'", v.namespace, clientCertificateAuth.SecretName)
+		}
+		key, ok := secret.Data["tls.key"]
+		if !ok {
+			return "", fmt.Errorf("no data for tls.key in secret '%s/%s'", v.namespace, clientCertificateAuth.SecretName)
+		}
+
+		clientCertificate, err := tls.X509KeyPair(cert, key)
+		if err != nil {
+			return "", fmt.Errorf("error reading client certificate: %s", err.Error())
+		}
+
+		// Setting up a short lived client with a configured client certificate.
+		// It is only meant to be used for requesting a Vault token. We clone
+		// http.Client's Transport seperately as it has to be adjusted and does
+		// not seem to be cloned by CloneConfig.
+		cfg := client.CloneConfig()
+		tmpTransport := cfg.HttpClient.Transport.(*http.Transport).Clone()
+		tmpTransport.TLSClientConfig.Certificates = append(tmpTransport.TLSClientConfig.Certificates, clientCertificate)
+		cfg.HttpClient.Transport = tmpTransport
+		client, err = vault.NewClient(cfg)
+		if err != nil {
+			return "", fmt.Errorf("error initializing intermediary Vault client: %s", err.Error())
+		}
+	}
+
+	parameters := map[string]string{
+		"role": clientCertificateAuth.Role,
```

Comment:
Seems like I assumed that "role" is a required parameter like it is for [Kubernetes auth](https://developer.hashicorp.com/vault/api-docs/auth/kubernetes#login).

However, for [client cert auth](https://developer.hashicorp.com/vault/api-docs/auth/cert#login-with-tls-certificate-method), only "name" is a valid parameter according to the documentation.

I guess this did not come up in my tests, as `name` - according to docs - is optional and defaults to returning any matching role, making my setup work, as there indeed was a single matching role.

Reply:
I renamed this parameter to match vault API as much as possible; Also made it optional, as it is optional for vault.